### PR TITLE
rewrite bootstrapping stages

### PR DESCRIPTION
**Pull request by mightyiam**
Continuing from #1327.

### CODE REVIEW

**Comment by mightyiam** [comment]
I did put in time and thought and feeling into this rewrite and I feel it's an improvement with regards to passing of information and use of language. Yet, it should go technical review, because I'm not savvy in most of what is involved.

**Comment by JohnTitor** [comment]
r? @jyn514 @tshepang maybe?
I don't feel I'm the best reviewer here.

File: src/building/bootstrapping.md [hunk 1]
```diff
@@ -47,54 +62,59 @@ graph TD
...
 
-### Stage 3
+#### Stage 3: the same-result test
 
-Stage 3 is optional. To sanity check our new compiler, we
-can build the libraries with the stage2 compiler. The result ought
-to be identical to before, unless something has broken.
+To verify that the stage 2 libraries that were copied from stage 1 are indeed
+identical to those which would otherwise have been compiled in stage 2, the
+stage 2 compiler is used to compile them and a comparison is made.
```
Comment (JohnTitor) [inline src/building/bootstrapping.md:113]: I guess emphasizing that stage 3 is optional is important here as most contributors don't need it.

Comment (mightyiam) [reply src/building/bootstrapping.md:113]: What is it, more specifically, that would require one to compile or use this stage compiler?

**Comment by jyn514** [comment]
@JohnTitor said this on the other pull request, but I'll repeat it here:

> > It was written carefully. I'm certain that it's a significant improvement.

> Could you elaborate on what improvements your changes are bringing? It's a bit unclear how these changes improve stuff.

I'm sorry you put in a lot of effort to this work, but it's a lot of effort to review as well, and it's unclear how it will help people reading the guide.

**Comment by mightyiam** [comment]
It should be clear to a reviewer how it will help people reading the guide.

**Comment by mightyiam** [comment]
Is there really nothing I can do to have this reviewed? Pinging someone specific? Asking for a budget for contracting a documentation expert?

**Comment by tshepang** [comment]
will have a look within the next week

File: src/building/bootstrapping.md [hunk 2]
```diff
@@ -2,24 +2,39 @@
...
+Each stage involves:
+- An existing compiler and its set of dependencies.
+- Targets ([objects][objects]): `std` and `rustc`.
 
-Compiling `rustc` is done in stages. Here's a diagram, adapted from Joshua Nelson's
-[talk on bootstrapping][rustconf22-talk] at RustConf 2022, with detailed explanations below.
+Note: the compiler of a stage—e.g. "the stage 1 compiler"—refers to the
+compiler that is compiled at that stage, not the one that already exists.
+
+In the first stage (stage 0) the compiler is usually obtained by downloading a
```
Comment (tshepang) [inline src/building/bootstrapping.md:32]: *usually* makes me wonder if other ways should be mentioned or linked to

File: src/building/bootstrapping.md [hunk 3]
```diff
@@ -47,54 +62,59 @@ graph TD
...
-When compiling `rustc`, the stage0 compiler uses the freshly compiled `std`.
-There are two concepts at play here:
-a compiler (with its set of dependencies)
-and its 'target' or 'object' libraries (`std` and `rustc`).
-Both are staged, but in a staggered manner.
+#### Stage 1: from current code, by an earlier compiler
 
-### Stage 1
+The stage 0 compiler compiles from current code `rustbuild` and `std` and uses
+them to compile from current code a compiler. This is the stage 1 compiler.
```
Comment (tshepang) [inline src/building/bootstrapping.md:77]: *produce* (or an equivalent) is an important word to keep here... makes it more clear what is happenin

File: src/building/bootstrapping.md [hunk 4]
```diff
@@ -47,54 +62,59 @@ graph TD
...
+expected to be identical.
 
-In theory, the stage1 compiler is functionally identical to the stage2 compiler,
-but in practice there are subtle differences.
-In particular, the stage1 compiler itself was built by stage0
-and hence not by the source in your working directory.
-This means that the symbol names used in the compiler source
-may not match the symbol names that would have been made by the stage1 compiler,
-which can cause problems for dynamic libraries and tests.
+The stage 1 compiler is used to compile from current code a compiler. This is
```
Comment (tshepang) [inline src/building/bootstrapping.md:88]: *produce* (or equivalent) also missing here

File: src/building/bootstrapping.md [hunk 5]
```diff
@@ -47,54 +62,59 @@ graph TD
...
+
+The symbol names used in the compiler source may not match the symbol names
+that would have been made by the stage1 compiler. This is important when using
+dynamic linking and due to the lack of ABI compatibility between versions. This
+primarily manifests when tests try to link with any of the `rustc_*` crates or
+use the (now deprecated) plugin infrastructure. These tests are marked with
+`ignore-stage1`.
+
+Also, the stage 2 compiler benefits from the compile-time optimizations
+generated by a compiler that is of the current code.
```
Comment (tshepang) [inline src/building/bootstrapping.md:107]: I would make these 2 bullet points, to show that they flow from "... differences"

**Comment by mightyiam** [comment]
Thanks a lot @tshepang! I will amend.

**Comment by JohnTitor** [comment]
Triage: Friendly-ping @mightyiam, is there any progress since your last comment?

**Comment by mightyiam** [comment]
Sorry. Thank you, @JohnTitor. Will get back to this.

**Comment by mightyiam** [comment]
Scheduled for tomorrow.

File: src/building/bootstrapping.md (hunk 1 above)
Comment (mightyiam) [reply src/building/bootstrapping.md]: Please determine whether it is resolved in HEAD.

File: src/building/bootstrapping.md [hunk 6]
```diff
@@ -2,24 +2,39 @@
...
 
-## Stages of bootstrapping
+Each stage involves:
+- An existing compiler and its set of dependencies.
+- Targets ([objects][objects]): `std` and `rustc`.
 
-Compiling `rustc` is done in stages. Here's a diagram, adapted from Joshua Nelson's
-[talk on bootstrapping][rustconf22-talk] at RustConf 2022, with detailed explanations below.
+Note: the compiler of a stage—e.g. "the stage 1 compiler"—refers to the
+compiler that is compiled at that stage, not the one that already exists.
```
[the remark this answers is not in the document]
Comment (mightyiam) [reply src/building/bootstrapping.md]: Resolving in favor of https://github.com/rust-lang/rustc-dev-guide/pull/1458#discussion_r1048000762.

File: src/building/bootstrapping.md (hunk 2 above)
Comment (mightyiam) [reply src/building/bootstrapping.md]: I agree. But I'm not familiar with other ways. And this doesn't seem like an issue created in this PR. Also, I've changed it to "typically". Resolve?

File: src/building/bootstrapping.md (hunk 3 above)
Comment (mightyiam) [reply src/building/bootstrapping.md]: Resolve in favor of https://github.com/rust-lang/rustc-dev-guide/pull/1458#discussion_r1048000762 ?

File: src/building/bootstrapping.md (hunk 4 above)
Comment (mightyiam) [reply src/building/bootstrapping.md]: Resolve in favor of https://github.com/rust-lang/rustc-dev-guide/pull/1458#discussion_r1048000762 ?

File: src/building/bootstrapping.md (hunk 5 above)
Comment (mightyiam) [reply src/building/bootstrapping.md]: Done. Resolve?

**Comment by mightyiam** [comment]
Sorry about my confusion regarding requesting of a review. Apparently, requesting review from one person removes the request for review from another... Anyway, @tshepang , please review.

**Comment by tshepang** [comment]
@mightyiam sorry to make you sad given your efforts and leading you on (by suggesting improvements you have now accepted), but this change does not feel like an overall improvement to existing text.

**Comment by tshepang** [comment]
The one change that may be worth submitting separately as a PR is the following, as it adds new info (but I would not be a proper reviewer for it as am not familiar)...
> This primarily manifests when tests try to link with any of the `rustc_*` crates or
  use the (now deprecated) plugin infrastructure. These tests are marked with
  `ignore-stage1`.

**Comment by mightyiam** [comment]
No, I'm sorry. Please provide feedback. Specific feedback, @tshepang. I am certain that this is an improvement.

**Comment by tshepang** [comment]
One example is you are saying the same thing, but making it more verbose, and less readable...

> `source code is compiled with the stage0 compiler` -> `stage 0 compiler produces from current code`

The explicit mention of current code feels overkill since no one is talking about old code.

One thing you may also submit as a separate PR, if not mentioned elsewhere, is...
> Note: the compiler of a stage—e.g. "the stage 1 compiler"—refers to the compiler that is produced at that stage, not the one that already exists.

**Comment by mightyiam** [comment]
> One example is you are saying the same thing, but making it more verbose, and/or less readable...
> 
> > `source code is compiled with the stage0 compiler` -> `stage 0 compiler produces from current code`
> 
> The explicit mention of current code feels overkill since no one is talking about old code.

The existence of a compiler that is from earlier code is central to the concept of bootstrapping.

What I am doing is giving each stage a title:

> - Stage 0: the pre-compiled compiler
> - Stage 1: from current code, by an earlier compiler
> - Stage 2: the truly current compiler
> - Stage 3: the same-result test

The phrasing "current code" originates in this context. And in this context I feel it is useful.

I have amended that sentence to include commas:

> The stage 0 compiler produces, from current code, `src/bootstrap` and `std` and uses them to produce, from current code, a compiler. This is the stage 1 compiler.

**Comment by jyn514** [comment]
We are not obligated to merge your code. The time you have spent so far has no impact on whether or not we merge your code. The only reason we would accept this change is because we think it is an improvement, not because you've worked hard on it.

**Comment by tshepang** [comment]
@mightyiam I hope this does not discourage you... feel free to submit additions/improvements, and try to keep them small, as that will make them more easy to review, and provide better feedback in case of disagreement

**Comment by mightyiam** [comment]
@tshepang and @jyn514, I am not discouraged. I certain that this PR is a significant improvement. I would like it reopened, please.

**Comment by mightyiam** [comment]
Documentation rewrites are sometimes the right thing to do. Are they impossible to review? This is not _that_ long. I dispute the rejection of it.

**Comment by jyn514** [comment]
> @JohnTitor said this on the other pull request, but I'll repeat it here:
> 
> > > It was written carefully. I'm certain that it's a significant improvement.
> 
> > Could you elaborate on what improvements your changes are bringing? It's a bit unclear how these changes improve stuff.
> 
> I'm sorry you put in a lot of effort to this work, but it's a lot of effort to review as well, and it's unclear how it will help people reading the guide.

@mightyiam you *still* 6 months later have not answered this question.

**Comment by jyn514** [comment]
The burden of proof is on you to show that it is an improvement, not on us to find fault with it.

**Pull request closed without merging**